src/Microsoft.OData.Core/ODataBatchReader.cs

```diff
@@ -204,48 +247,167 @@ void IODataBatchOperationListener.BatchOperationContentStreamDisposed()
+        /// in a batch request.
+        /// </summary>
+        /// <returns>The message that can be used to read the content of the batch request operation from.</returns>
+        protected abstract ODataBatchOperationRequestMessage CreateOperationRequestMessageImplementation();
+
+        /// <summary>
+        /// Returns the cached <see cref="ODataBatchOperationRequestMessage"/> for reading the content of an operation
+        /// in a batch request.
+        /// </summary>
+        /// <returns>The message that can be used to read the content of the batch request operation from.</returns>
+        protected abstract ODataBatchOperationResponseMessage CreateOperationResponseMessageImplementation();
+
+        /// <summary>
+        /// Implementation of the reader logic when in state 'Start'.
+        /// </summary>
+        /// <returns>The batch reader state after the read.</returns>
+        protected abstract ODataBatchReaderState ReadAtStartImplementation();
+
+        /// <summary>
+        /// Implementation of the reader logic when in state 'Operation'.
+        /// </summary>
+        /// <returns>The batch reader state after the read.</returns>
+        protected abstract ODataBatchReaderState ReadAtOperationImplementation();
+
+        /// <summary>
+        /// Implementation of the reader logic when in state 'ChangesetStart'.
+        /// </summary>
+        /// <returns>The batch reader state after the read.</returns>
+        protected abstract ODataBatchReaderState ReadAtChangesetStartImplementation();
+
+        /// <summary>
+        /// Implementation of the reader logic when in state 'ChangesetEnd'.
+        /// </summary>
+        /// <returns>The batch reader state after the read.</returns>
+        protected abstract ODataBatchReaderState ReadAtChangesetEndImplementation();
+
+        /// <summary>
+        /// Instantiate an <see cref="ODataBatchOperationRequestMessage"/> instance.
+        /// </summary>
+        /// <param name="streamCreatorFunc">The function for stream creation.</param>
+        /// <param name="method">The HTTP method used for this request message.</param>
+        /// <param name="requestUri">The request Url for this request message.</param>
+        /// <param name="headers">The headers for this request message.</param>
+        /// <param name="contentId">The contentId of this request message.</param>
+        /// <param name="dependsOnRequestIds">
+        /// The prerequisite request Ids of this request message. For batch in Json format,
+        /// some of these request Ids are resolved from prerequisite atomic groups that are
+        /// specified in the dependsOn attribute of the request.</param>
+        /// <returns>The <see cref="ODataBatchOperationRequestMessage"/> instance.</returns>
+        protected ODataBatchOperationRequestMessage BuildOperationRequestMessage(

```

Comment:
>BuildOperationRequestMessage [](start = 52, length = 28)

Why doesn't buildoperationrequestmessage in batchreader set the groupid on the request message?  Given that it can be set when creating a request message, why wouldn't we set it when reading the request message?

Reply:
Moved the setting groupid here.

---
In reply to: [155353014](https://github.com/OData/odata.net/pull/992#discussion_r155353014) [](ancestors = 155353014)